### PR TITLE
Multiple ghc-9.4.x build fixes

### PR DESCRIPTION
**Pull request by peti**
All changes tested and verified on `x86_64-linux`.

### CODE REVIEW

**Comment by sternenseemann** [comment]
Rebased and switched target to `haskell-updates`, made commit messages conformant with CONTRIBUTING.md while I was at it.

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix [hunk 1]
```diff
@@ -184,22 +184,38 @@ in {
...
+    hls-floskell-plugin = null;   # This plugin is supposed to work, but fails to compile.
+    hls-rename-plugin = null;     # This plugin is supposed to work, but fails to compile.
     hls-stylish-haskell-plugin = null;
   };
 
   # https://github.com/tweag/ormolu/issues/941
-  ormolu = overrideCabal (drv: {
-    libraryHaskellDepends = drv.libraryHaskellDepends ++ [ self.file-embed ];
-  }) (disableCabalFlag "fixity-th" super.ormolu);
+  ormolu = self.ormolu_0_5_2_0;
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:198]:
```suggestion
  ormolu = doDistribute self.ormolu_0_5_2_0;
```

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix [hunk 2]
```diff
@@ -184,22 +184,38 @@ in {
...
   fourmolu = overrideCabal (drv: {
     libraryHaskellDepends = drv.libraryHaskellDepends ++ [ self.file-embed ];
   }) (disableCabalFlag "fixity-th" super.fourmolu_0_10_1_0);
+
+  # The Haskell library has additional dependencies when compiled with ghc-9.4.x.
+  X11-xft = addExtraLibraries [pkgs.xorg.libXau pkgs.xorg.libXdmcp pkgs.expat] super.X11-xft;
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:204]: This is due to https://github.com/haskell/cabal/issues/8455. This override should not be required, as it is clearly a Cabal bug. I would prefer that we not add such workarounds yet, hoping that a patch for the problem will become available.

Another possibility could be using Cabal 3.6 for building `Setup.hs`.

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix [hunk 3]
```diff
@@ -184,22 +184,38 @@ in {
...
+
+  # The gtk2hs setup hook provided by this package lacks the ppOrdering field that
+  # recent versions of Cabal require. This leads to builds like cairo and glib
+  # failing during the Setup.hs phase: https://github.com/gtk2hs/gtk2hs/issues/323.
+  gtk2hs-buildtools = appendPatch ./patches/gtk2hs-buildtools-fix-ghc-9.4.x.patch super.gtk2hs-buildtools;
+
+  # Package 'libpcre2-8', required by 'glib-2.0', not found.
+  # text >=1.0.0.0 && <1.3.
+  glib = addExtraLibrary pkgs.pcre2 (doJailbreak super.glib);
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:213]: Same note about https://github.com/haskell/cabal/issues/8455.

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix [hunk 4]
```diff
@@ -184,22 +184,38 @@ in {
...
+  cairo = addExtraLibraries [pkgs.pcre2 pkgs.xorg.libXdmcp] (doJailbreak super.cairo);
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:214]: Same note about https://github.com/haskell/cabal/issues/8455.

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix [hunk 5]
```diff
@@ -184,22 +184,38 @@ in {
...
+
+  # In addition to the overrides required by glib and cairo, pango needs the
+  # pkg-config file for mount.pc, which in turn requires a whole lot of other
+  # pkg-config files.
+  pango = addExtraLibraries (with pkgs; [pcre2 xorg.libXdmcp util-linuxMinimal libselinux libsepol pcre fribidi libthai libdatrie]) (doJailbreak super.pango);
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:219]: Same note about https://github.com/haskell/cabal/issues/8455.

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix (hunk 2 above)
Comment (peti) [reply pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:204]: I would like to use GHC 9.4.4 *now*. Hoping that a patch might emerge on the future doesn't really help me much with that goal.

**Comment by peti** [comment]
@sternenseemann , I am not sure what you would like me to do. Are you opposed to merging those fixes?

**Comment by sternenseemann** [comment]
See https://github.com/NixOS/nixpkgs/pull/214446.

**Comment by peti** [comment]
> See #214446.

👍

**Pull request merged**